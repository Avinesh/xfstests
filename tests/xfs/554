#! /bin/bash
# SPDX-License-Identifier: GPL-2.0
# Copyright (c) 2022 Fujitsu Limited. All Rights Reserved.
#
# FS QA Test No. 554
#
# Create a filesystem which contains an inode with a lower number
# than the root inode. Set the lower number to a dump file as the root inode
# and ensure that 'xfsrestore -x' handles this wrong inode.
#
. ./common/preamble
_begin_fstest auto quick dump

# Import common functions.
. ./common/dump

_supported_fs xfs
_fixed_by_git_commit xfsdump \
	"XXXXXXXXXXXX xfsrestore: fix rootdir due to xfsdump bulkstat misuse"
_require_xfs_io_command "falloc"
_require_scratch
_require_xfsrestore_xflag

# A large stripe unit will put the root inode out quite far
# due to alignment, leaving free blocks ahead of it.
_scratch_mkfs_xfs -d sunit=1024,swidth=1024 > $seqres.full 2>&1 || _fail "mkfs failed"

# Mounting /without/ a stripe should allow inodes to be allocated
# in lower free blocks, without the stripe alignment.
_scratch_mount -o sunit=0,swidth=0

root_inum=$(stat -c %i $SCRATCH_MNT)

# Consume space after the root inode so that the blocks before
# root look "close" for the next inode chunk allocation
$XFS_IO_PROG -f -c "falloc 0 16m" $SCRATCH_MNT/fillfile

# And make a bunch of inodes until we (hopefully) get one lower
# than root, in a new inode chunk.
echo "root_inum: $root_inum" >> $seqres.full
for i in $(seq 0 4096) ; do
	fname=$SCRATCH_MNT/$(printf "FILE_%03d" $i)
	touch $fname
	inum=$(stat -c "%i" $fname)
	[[ $inum -lt $root_inum ]] && break
done

echo "created: $inum" >> $seqres.full

[[ $inum -lt $root_inum ]] || _notrun "Could not set up test"

# Now try a dump and restore. Cribbed from xfs/068
_create_dumpdir_stress

echo -n "Before: " >> $seqres.full
_count_dumpdir_files | tee $tmp.before >> $seqres.full

_do_dump_file

# Set the wrong root inode number to the dump file
# as problematic xfsdump used to do.
$here/src/fake-dump-rootino $dump_file $inum

_do_restore_file -x | \
sed -e "s/rootino #${inum}/rootino #FAKENO/g" \
	-e "s/# to ${root_inum}/# to ROOTNO/g" \
	-e "/entries processed$/s/[0-9][0-9]*/NUM/g"

echo -n "After: " >> $seqres.full
_count_restoredir_files | tee $tmp.after >> $seqres.full
diff -u $tmp.before $tmp.after

# success, all done
status=0
exit
