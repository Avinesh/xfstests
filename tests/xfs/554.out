QA output created by 554
Creating directory system to dump using fsstress.

-----------------------------------------------
fsstress : -f link=10 -f creat=10 -f mkdir=10 -f truncate=5 -f symlink=10
-----------------------------------------------
Dumping to file...
xfsdump  -f DUMP_FILE -M stress_tape_media -L stress_554 SCRATCH_MNT
xfsdump: using file dump (drive_simple) strategy
xfsdump: level 0 dump of HOSTNAME:SCRATCH_MNT
xfsdump: dump date: DATE
xfsdump: session id: ID
xfsdump: session label: "stress_554"
xfsdump: ino map <PHASES>
xfsdump: ino map construction complete
xfsdump: estimated dump size: NUM bytes
xfsdump: /var/xfsdump/inventory created
xfsdump: creating dump session media file 0 (media 0, file 0)
xfsdump: dumping ino map
xfsdump: dumping directories
xfsdump: dumping non-directory files
xfsdump: ending media file
xfsdump: media file size NUM bytes
xfsdump: dump size (non-dir files) : NUM bytes
xfsdump: dump complete: SECS seconds elapsed
xfsdump: Dump Status: SUCCESS
Restoring from file...
xfsrestore  -x -f DUMP_FILE  -L stress_554 RESTORE_DIR
xfsrestore: using file dump (drive_simple) strategy
xfsrestore: using online session inventory
xfsrestore: searching media for directory dump
xfsrestore: examining media file 0
xfsrestore: reading directories
xfsrestore: found fake rootino #FAKENO, will fix.
xfsrestore: fix root # to ROOTNO (bind mount?)
xfsrestore: NUM directories and NUM entries processed
xfsrestore: directory post-processing
xfsrestore: restoring non-directory files
xfsrestore: restore complete: SECS seconds elapsed
xfsrestore: Restore Status: SUCCESS
